### PR TITLE
Dependency injection container

### PR DESCRIPTION
**Pull request by CarolinWickner**
- Moved controller and repository to dependency injection container
- Improved validation when adding a new project
- Refactored controller, service and repository

### CODE REVIEW

File: src/AppBundle/Service/StaticDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        $this->add('Awesome Project', $startDate1, $endDate1, 'Dont forget to be awesome.');
+        $this->add('Another Awesome Project', $startDate2, $endDate2, 'Dont forget to be awesome.');
+    }
+
+    public function get() : array 
+    {  
+        return  $this->projects;
+    }
+    
+    public function add(string $name, \DateTime  $startDate, \DateTime  $endDate, string $teaser) : Project
```
Comment (mfranczak) [inline src/AppBundle/Service/StaticDataProvider.php]: I think your code throws an error that add method is not compatible with DataProvider interface.

File: src/AppBundle/Service/StorageDataProvider.php [hunk 2]
```diff
@@ -17,6 +20,11 @@ class ProjectRepository {
     */
     private $nextId;
 
+    public function __construct($validator)
```
Comment (mfranczak) [inline src/AppBundle/Service/StorageDataProvider.php]: Just for the sake of discussion. Classes should have only one reason for change. Your repository at this point of time is responsible for:
- validation
- holding entities
- object construction this is way to much
- managing ids

Do you have an idea how could you potentially restructure it?

File: src/AppBundle/Service/DataProvider.php [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+declare(strict_types = 1);
+
+namespace AppBundle\Service;
+
+interface DataProvider
+{
+    public function get();
+
+    public function add(string $name, \DateTime  $startDate, \DateTime  $endDate, string $teaser);
```
Comment (mfranczak) [inline src/AppBundle/Service/DataProvider.php]: Consider the following change request:
"We want our Project to have new fields: owner, description, icon, creation date, technologies"

With your implementation you would change the signature of this method and the change in each repository the way you construct the object. 

Do you have an idea how to make your repository implementation more flexible?

File: src/AppBundle/Service/StaticDataProvider.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        $startDate1 = new \DateTime('2019-01-05');
+        $endDate1 = new \DateTime('2019-06-05');
+        $startDate2 = new \DateTime('2019-01-05');
+        $endDate2 = new \DateTime('2019-06-05');
+
+        $this->add('Awesome Project', $startDate1, $endDate1, 'Dont forget to be awesome.');
+        $this->add('Another Awesome Project', $startDate2, $endDate2, 'Dont forget to be awesome.');
+    }
+
+    public function get() : array
```
Comment (mfranczak) [inline src/AppBundle/Service/StaticDataProvider.php]: It breaks the interface.

File: src/AppBundle/Controller/ProjectController.php [hunk 5]
```diff
@@ -4,53 +4,58 @@
...
+        try { 
+            $project = $this->provider->add(
+                $data['name'], 
+                new \DateTime($data['startDate']),
+                new \DateTime($data['endDate']),
+                $data['teaser']
+            );
+            $response = new JsonResponse(
+                $project,
+                Response::HTTP_CREATED
```
Comment (mfranczak) [inline src/AppBundle/Controller/ProjectController.php]: Nice!

File: src/AppBundle/Controller/ProjectController.php [hunk 6]
```diff
@@ -4,53 +4,58 @@
...
+            );
+        } 
+        catch (ValidationException $e) {
             $response = new Response(
-                json_encode($project),
-                Response::HTTP_CREATED,
-                ['content-type' => 'application/json']
+                $e->getConstraintViolationList(),
+                Response::HTTP_BAD_REQUEST
```
Comment (mfranczak) [inline src/AppBundle/Controller/ProjectController.php]: Nice!

File: src/AppBundle/Service/StorageDataProvider.php [hunk 7]
```diff
@@ -52,8 +60,14 @@ public function add(string $name, \DateTime  $startDate, \DateTime  $endDate, st
         $projects = $this->get();
         $id = $this->nextId++;
         $newProject = new Project($id, $name, $startDate, $endDate, $teaser);
+
+        $errors = $this->validator->validate($newProject);
+        if (count($errors) > 0) {
+            throw new ValidationException($errors);
+        } else {
         $projects[] = $newProject;
         file_put_contents(__DIR__ . '/../storage.json', json_encode($projects));
```
Comment (mfranczak) [inline src/AppBundle/Service/StorageDataProvider.php]: It's a good practice to check if the operation succeed or not. If not throw an exception.

File: src/AppBundle/Service/StaticDataProvider.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    */
+    private $projects;
+
+    /**
+    * @var int
+    */
+
+    private $nextId;
+
+    public function __construct($validator)
```
Comment (mfranczak) [inline src/AppBundle/Service/StaticDataProvider.php]: Can you try to figure out what is the interface of this class to use the correct type for this argument.

File: src/AppBundle/Service/DataProvider.php [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AppBundle\Service;
+
+interface DataProvider
```
Comment (mfranczak) [inline src/AppBundle/Service/DataProvider.php]: Make sure that the return types are correct

Comment (mfranczak) [reply src/AppBundle/Service/DataProvider.php]: And why not repository?

File: src/AppBundle/Service/StorageDataProvider.php (hunk 2 above)
Comment (CarolinWickner) [reply src/AppBundle/Service/StorageDataProvider.php]: The business logic part (constructing objects and validation) should probably be a service whereas holding entities and managing id's should be in repository.

**Pull request merged**